cmd/root.go

```diff
@@ -51,6 +60,88 @@ func init() {
 	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
 }
 
+// isArgTuneable returns if the argument provided is allowed to be modified
+func isArgTunable(arg string) bool {
+	return tuneableArgsWhitelist[arg]
+}
+
+// parseTuningFile parses the kernel argument tuning file
+func parseTuningFile(tuningFilePath string) (addArguments []types.TuneArgument, deleteArguments []types.TuneArgument, err error) {
+	if tuningFilePath == "" {
+		tuningFilePath = kernelTuningFile
+	}
+	// Return fast if the file does not exist
+	if _, err := os.Stat(tuningFilePath); os.IsNotExist(err) {
+		glog.V(2).Infof("no kernel tuning needed as %s does not exist", tuningFilePath)
+		// This isn't an error. Return out.
+		return addArguments, deleteArguments, err
+	}
+	// Read and parse the file
+	file, err := os.Open(tuningFilePath)
+	if err != nil {
+		// If we have an issue reading return an error
+		glog.Infof("Unable to open %s for reading: %v", tuningFilePath, err)
+		return addArguments, deleteArguments, err
+	}
+
+	// Parse the tuning lines
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, "ADD") {
+			// NOTE: Today only specific bare kernel arguments are allowed so
+			// there is not a need to split on =.
+			key := line[4:]
+			if isArgTunable(key) {
+				addArguments = append(addArguments, types.TuneArgument{Key: key, Bare: true})
+			} else {
+				glog.Infof("%s not a whitelisted kernel argument", key)
+			}
+		} else if strings.HasPrefix(line, "DELETE") {
```

Comment:
I don't think we should ship `DELETE` functionality now.  I'm not aware of a use case for deleting our built-in kernel args, and this is going to be messy if anyone tries to use it to remove `nosmt` after a cluster has started.

(What we really want is idempotence; today e.g. `rpm-ostree kargs --delete nosmt` will work the *first* time and then fail subsequent times as the argument won't exist.  Which is right probably for a CLI, but a declarative system like MachineConfig would resolve this)

Reply:
Hm...actually the fact that pivot consumes the `/etc/pivot/kernel-args` (as a way to implement "at most once" semantics) means that it *must not* be part of the MachineConfig, or otherwise the MCD will go degraded as the file it expects to exist won't.

`/etc/pivot/kernel-args` must then be injected into the Ignition, not in the MachineConfig.

(All of this again is IMO a strong argument for having "add these kernel arguments" in pivot, and day 2 operations driven by the MCD)

Reply:
Now, we could also implement `rpm-ostree kargs --ensure --delete=nosmt --exit-unchanged-77` and `--ensure` here means "it's not an error if it doesn't exist".  I think we should do that, as it would make it easier for the MCD and pivot to implement idempotent sync.

(Imagine for example that RHCOS eventually defaults to `nosmt`, then `rpm-ostree kargs --ensure --add=nosmt` would become a no-op)

Reply:
We will need to be able to remove the whitelisted parameters. However, I can add a check to not run the command if said parameter is not there.

Reply:
I don't understand how this could work.  I'm a cluster admin, day 2 I decide I didn't want `nosmt` I.e. I choose to enable perf over security.

How do I use this `DELETE` bit?

Reply:
(Because if you're thinking "create a MC which writes `/etc/pivot/kernel-args` that's going to break per my earlier comment)

Reply:
In other words, I'm not disagreeing there's a need - there's obviously a need.  But I don't think we should implement it in pivot.

Reply:
@cgwalters ok, fair. Where do you see this being implemented?

Reply:
My vote is we just ship this "add nosmt in early pivot", and then the answer for this "day 2 I decide to enable SMT" would probably be:

```
for node in nodes:
  oc debug $node chroot /host rpm-ostree kargs --remove=nosmt
```

And then create any dummy MC file to cause the MCD to reboot.

I don't expect many people will need the ability to remove `nosmt`.

Reply:
Removed DELETE functionality in a fixup.